src/tribler-core/tribler_core/components/tag/community/tag_community.py

```diff
@@ -0,0 +1,54 @@
+import json
+import random
+from binascii import unhexlify
+
+from pony.orm import TransactionIntegrityError, db_session
+
+from ipv8.lazy_community import lazy_wrapper
+from tribler_core.components.metadata_store.db.store import MetadataStore
+from tribler_core.components.tag.community.tag_payload import TagMessage, TagModel
+from tribler_core.modules.tribler_community import TriblerCommunity
+
+GOSSIP_RANDOM_TAGS_COUNT = 10
+GOSSIP_RANDOM_PEERS_COUNT = 10
+
+
+class TagCommunity(TriblerCommunity):
+    """
+    Community for disseminating tags across the network.
+    """
+
+    community_id = unhexlify('042020c5e5e2ee0727fe99d704b430698e308d98')
+
+    def __init__(self, *args, metadata_store: MetadataStore = None,
+                 gossip_interval_for_tags_in_sec=5, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.metadata_store = metadata_store
+
+        self.add_message_handler(TagMessage, self.on_message)
+        self.register_task("gossip_random_tags", self.gossip_random_tags, interval=gossip_interval_for_tags_in_sec)
+
+        self.logger.info('Tag community initialized')
+
+    @lazy_wrapper(TagMessage)
+    def on_message(self, peer, payload):
+        self.logger.debug(f'Message received: {payload}')
+        dictionary = json.loads(payload.json.decode("utf-8"))
+        tag_model = TagModel.parse_obj(dictionary)
+        self.logger.debug(f'Write to db: {tag_model}')
+        try:
+            with db_session():
+                self.metadata_store.Tag(**tag_model.to_pony_dict())
+        except TransactionIntegrityError:  # ignore all duplicates
+            pass
+
+    def gossip_random_tags(self):
```

Comment:
**Suggestion**: consider using pull-based gossip to prevent the simplest forms of spam attacks.

Reply:
I remember that sentence of yours.
However, after reasoning, I have come to the conclusion that in terms of "spam attacks"  it doesn't matter whether we use push or pull.

Both of these approaches put approximately the same load on the lower layer protocol (ipv8). We can filter spam messages only on the Community level.

Pull-based gossip will lead to the use of a combination of a dictionary (for storing outcoming requests) and a `NumberCache`.

Push-based gossip needs neither.

To add simple prevention of "spam attack" for "push-based" gossip we just may control the frequency of incoming messages and drop all messages that come frequently than it is accepted (or peers who have exceeded the limit).  

We also add a filter for the GUI interface "only show a tag if it has been suggested by two distinct identities".

From my POV it should be enough for the first version of Tag System.

That's why I choose push-based gossip. It gives roughly the same level of spam protection, but it is simplier.

Reply:
> It gives roughly the same level of spam protection

No, it definitely does not give the same level of protection, unless you add lots of things like controlling the rate of incoming packets - which defies the whole point of simplicity. 

It will take one mischievous student to take down the whole network in flames, if you use **both** blind transitivity **and** push-based gossip, even if you add rate control.

Anyways, that's your choice. Let's just hope you're still with us when :hankey:  hits the fan :wink:

Reply:
For comparison, we are using pull-based gossip in the `BandwidthAccountingCommunity`, see [here](https://github.com/Tribler/tribler/blob/main/src/tribler-core/tribler_core/components/bandwidth_accounting/community/bandwidth_accounting_community.py#L145). Unfortunately, I don't exactly remember the reasoning behind that design 😕 . We can discuss it further on Monday.